### PR TITLE
[KYUUBI #751] Support single session mode

### PR DESCRIPTION
**Pull request by wForget**
### _Why are the changes needed?_

#751

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala [hunk 1]
```diff
@@ -53,7 +54,13 @@ class SparkSQLSessionManager private (name: String, spark: SparkSession)
     val sessionImpl = new SparkSessionImpl(protocol, user, password, ipAddress, conf, this)
     val handle = sessionImpl.handle
     try {
-      val sparkSession = spark.newSession()
+      val sparkSession = if (spark.sessionState.conf
+        .getConf(StaticSQLConf.HIVE_THRIFT_SERVER_SINGLESESSION)) {
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: Can we add a new config in `KyuubiConf` ? We don't want to introduce any thriftserver config.

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/session/SingleSessionSuite.scala [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.engine.spark.session
+
+import org.apache.kyuubi.config.KyuubiConf.ENGINE_SHARE_LEVEL
+
+class SingleSessionSuite extends SessionSuite{
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/session/SingleSessionSuite.scala:22]: nit: `SessionSuite  {`

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/session/SingleSessionSuite.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    withJdbcStatement() { statement =>
+      statement.execute("create or replace temporary view temp_view as select 1")
+    }
+    withJdbcStatement() { statement =>
+      val resultSet = statement.executeQuery("select * from temp_view")
+      assert(resultSet.next())
+      assert(resultSet.getInt(1) == 1)
+    }
+  }
+
```
Comment (ulysses-you) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/session/SingleSessionSuite.scala:39]: nit: unnecessary line

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala (hunk 1 above)
Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: And don't forget to generate the docs with cmd
`KYUUBI_UPDATE=1 build/mvn test -DwildcardSuites=org.apache.kyuubi.config.AllKyuubiConfiguration`

**Review by ulysses-you — COMMENTED** [review]
thank you for making this PR, welcome to Kyuubi !

**Comment by wForget** [comment]
> thank you for making this PR, welcome to Kyuubi !

Thank you for your reply, I will improve it according to your suggestions.

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala (hunk 1 above)
Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: > Can we add a new config in `KyuubiConf` ? We don't want to introduce any thriftserver config.

Is it okay to use `kyuubi.session.engine.singleSession` configuration name?

Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: We can make it more simple `kyuubi.engine.singleton`

Comment (ulysses-you) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: FYI @turboFei @pan3793 @yaooqinn

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: kyuubi.engine.singleton looks overkill

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: how about `kyuubi.engine.single.spark.session`? `kyuubi.engine.singleton` seems to tell users that only one engine will be created

Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: +1 for `kyuubi.engine.single.spark.session`

Comment (wForget) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/kyuubi/engine/spark/session/SparkSQLSessionManager.scala:58]: `kyuubi.engine.single.spark.session` looks good

**Comment by wForget** [comment]
Hi @ulysses-you @yaooqinn @pan3793 .
Thanks, it has been improved according to your suggestions, please review it again.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
LGTM

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 4]
```diff
@@ -657,4 +657,13 @@ object KyuubiConf {
     .version("1.1.1")
     .stringConf
     .createOptional
+
+  val ENGINE_SINGLE_SESSION =
+    buildConf("engine.single.spark.session")
```
Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:662]: should we rename it to `kyuubi.engine.single.session`?

Comment (turboFei) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:662]:
```
val ENGINE_SINGLE_SESSION =
    buildConf("engine.single.session")
```
or 
```
val ENGINE_SINGLE_SPARK_SESSION =
    buildConf("engine.single.spark.session")
```
? cc @pan3793

Comment (turboFei) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:662]: hi @wForget  could you keep it consistent for the VARIABLE name and config entry name?

Comment (wForget) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:662]: Ok thanks, i will fix it

**Comment by yaooqinn** [comment]
thanks, merged to master

**Comment by ulysses-you** [comment]
late lgtm

**Pull request closed without merging**